# Nanopore data processing

Alignment and other processing steps for Nanopore data.

---

### Minimap2 and NGMLR to realign complex regions

Originally, we planned to use the Vulcan pipeline (https://gitlab.com/treangenlab/vulcan/-/tree/master) for alignment and realigning complex regions to get better SV calling downstream. But, Vulcan is not very customizable, so we'll follow some of their general steps (Minimap2 and Samtools steps) written as separate bash scripts combined with modifications to their `vulcan` script.

Align with Minimap2 and only keep primary mapping.

```bash
# In dir: ~/GitHub/Barley_Mutated/00_sequence_processing/Nanopore
sbatch np_read_mapping-Morex-sample2.job
# Keep primary mapping and add @SQ header lines
sbatch keep_primary_mapping-Morex-sample2.job

# Repeat for 3 mutated barley samples (M01, M20, M29)
sbatch np_read_mapping-M01.job
sbatch np_read_mapping-M20.job
sbatch np_read_mapping-M29.job
# Keep primary mapping and add @SQ header lines
# except submit as Slurm task array
sbatch --array=0-2 keep_primary_mapping-mut_lines.job
```

After modifying the `vulcan` script, transferred the script to MSI. The modification was adding the `--full_sam` and `--full_sam_primary` options to take a custom generated Minimap2 SAM files as input. Modifications are documented in this forked repository under [commit 0dec3094](https://gitlab.com/ChaochihL/vulcan/-/commit/8dce5d4eb75a6044e0fcc00894e22933f56e91c2). The modified `vulcan` script and documentation are available in this forked repository: https://gitlab.com/ChaochihL/vulcan.

```bash
# In dir: ~/GitHub/vulcan
scp vulcan liux1299@mesabi.msi.umn.edu:/panfs/roc/groups/9/morrellp/liux1299/.conda/envs/vulcan_env/bin
```

Run modified vulcan pipeline.

```bash
# In dir: ~/GitHub/Barley_Mutated/00_sequence_processing/Nanopore
sbatch vulcan_read_mapping-Morex-sample2_partsRef.job

# Repeat for 3 mutated barley samples (M01, M20, M29)
sbatch vulcan_read_mapping-M01_partsRef.job
sbatch vulcan_read_mapping-M20_partsRef.job
sbatch vulcan_read_mapping-M29_partsRef.job
```

The Vulcan pipeline outputs a sorted BAM file.

Add @RG header line to BAM file, many downstream programs require @RG header lines to be present.

```bash
# In dir: ~/GitHub/Barley_Mutated/00_sequence_processing/Nanopore
sbatch add_RG_header.sh

# Repeat for 3 mutated barley samples (M01, M20, M29)
sbatch add_RG_header-M01.sh
sbatch add_RG_header-M20.sh
sbatch add_RG_header-M29.sh
```

Run Sniffles2.

```bash
# In dir: ~/GitHub/Barley_Mutated/00_sequence_processing/Nanopore
sbatch sniffles-Morex-sample2_partsRef.job

# Repeat for 3 mutated barley samples (M01, M20, M29)
sbatch sniffles-M01.job
sbatch sniffles-M20.job
sbatch sniffles-M29.job
```

Reheader VCF so `SAMPLE` gets replaced with actual sample ID.

```bash
# In dir: ~/GitHub/Barley_Mutated/00_sequence_processing/Nanopore
./reheader_vcf-ONT_Morex-sample2.sh
```

Get high coverage regions.

```bash
# In dir: ~/GitHub/Barley_Mutated/00_sequence_processing/Nanopore
# Estimate coverage per window
sbatch --array=0 run_mosdepth-ont.sh

# Extract high coverage regions
sbatch run_extract_high_depth_regions-ont.sh

# Merge near regions
# In dir: ~/Alignments/nanopore_morex/Morex_ont_partsRefv3/per_sample_high_depth
module load bedtools/2.29.2
bedtools merge -d 10 -i Morex_ont_partsRefv3_90_wRG.high_cov.bed > Morex_ont_partsRefv3_90_wRG.high_cov.exclude.bed
```

Get coverage with labels to identify low coverage regions. The script also outputs low coverage regions for IGV exploration and possible filtering.

```bash
# In dir: ~/GitHub/Barley_Mutated/00_sequence_processing/Nanopore
sbatch mosdepth_coverage_labels-morex.sh
sbatch mosdepth_coverage_labels-M01.sh
sbatch mosdepth_coverage_labels-M20.sh
sbatch mosdepth_coverage_labels-M29.sh
```

Prepare low coverage regions output for IGV explorationa and possible filtering.

---

### Morex 85X ONT data

For exploratory purposes only, not included in final filtering of Morex.

BAM file was generated by Giulia and is located at:

```bash
/panfs/roc/groups/9/morrellp/gfrascar/Morex_IPK_85X/Morex_85X_sorted.bam
```

Run Sniffles2.

```bash
# In dir: ~/GitHub/Barley_Mutated/00_sequence_processing/Nanopore
# Repeat for 85X ONT Morex
sbatch sniffles-85X_ONT_Morex_pseudo.job
```

Reheader VCF so `SAMPLE` gets replaced with actual sample ID.

```bash
# In dir: ~/GitHub/Barley_Mutated/00_sequence_processing/Nanopore
./reheader_vcf-ONT_Morex_85x.sh
```

**Convert Sniffles VCF pseudo to parts positions:** Morex 85x ONT data was originally called using the barley pseudomolecules plastids reference fasta file, so we'll need to convert the positions to parts positions to compare with our other datasets. *IMPORTANT!* We'll need a special script to do the conversion since the end positions stored in the `INFO/END` field also need to be updated accordingly (CAN'T use the conversion script `~/GitHub/File_Conversions/Barley_Pseudomolecules_to_Parts.py`).

```bash
# Dependencies
module load python3/3.8.3_anaconda2020.07_mamba
module load htslib/1.9
PICARD_JAR="/panfs/roc/groups/9/morrellp/public/Software/picard_ML_2.20.2/picard.jar"

# User provided input arguments
VCF="/panfs/roc/groups/9/morrellp/shared/Datasets/Alignments/nanopore_morex/Morex_85x_ont/Morex_85X_sorted_renamed.vcf"
REF_DICT="/panfs/roc/groups/9/morrellp/shared/References/Reference_Sequences/Barley/Morex_v3/Barley_MorexV3_pseudomolecules_plastids_parts.dict"

# Sniffles pseudo to parts conversion
VCF_DIR=$(dirname ${VCF})
VCF_PREFIX=$(basename ${VCF} .vcf)
~/GitHub/Barley_Mutated/01_snp_filtering/sniffles_vcf_pseudo_to_parts.py ${VCF} > ${VCF_DIR}/temp_${VCF_PREFIX}_parts.vcf
bgzip ${VCF_DIR}/temp_${VCF_PREFIX}_parts.vcf
tabix -p vcf --csi ${VCF_DIR}/temp_${VCF_PREFIX}_parts.vcf.gz

# Fix chr and chr length in header lines using Picard
~/GitHub/morex_reference/morex_v2/50k_9k_BOPA_SNP/scripts/add_contig_length_to_header.sh \
    ${VCF_DIR}/temp_${VCF_PREFIX}_parts.vcf.gz \
    ${REF_DICT}

# Rename and cleanup intermediate files
mv ${VCF_DIR}/temp_${VCF_PREFIX}_parts_fixedHeader.vcf ${VCF_DIR}/${VCF_PREFIX}_parts.vcf
rm ${VCF_DIR}/temp_${VCF_PREFIX}_parts.vcf*
```
