# Barley_Mutated

This repository reports the scripts and code written in support of a manuscript by Chaochih Liu et al. examining the nature of mutations generated by sodium azide treatment in barley. Mutagen induced variants (SNPs and indels) are compared to non-mutagenized (naturally occurring) variants in barley landraces.

The primary data is derived from DNA whole genome resequencing, including Illumina paired-end reads, Illumina reads with a 10X Genomics linked read library preparation, and Oxford Nanopore (ONT) long reads. We also make use of previously published PacBio data.

## Analysis overview

Variant calling in this project makes use of GATK 4 to identify SNPs and indels in Illumina resequencing data. Structural variants were identified using multiple lines of evidence through the 10X Genomics linked reads technology and Sniffles2 and cuteSV for the ONT data.

The 10X Genomics linked reads were processed using [LongRanger](https://doi.org/10.1101/gr.234443.118). Oxford Nanopore sequencing was used for read-based verification of structural variants, with variant calls based on [Sniffles](https://doi.org/10.1038/s41592-018-0001-7) and [cuteSV](https://github.com/tjiangHIT/cuteSV) as implemented through [Vulcan](https://doi.org/10.1093/gigascience/giab063) with customizations made to handle large barley Nanopore data that are available in this forked repository: https://gitlab.com/ChaochihL/vulcan.

Careful filtering was done to minimize spurious calls and to identify variants induced by the mutagen. This step is important because characterizing mutagen induced variants involves multiple generations of seed bulking and selfing following the mutagen treatment. This means we're always evaluating a mixture of mutagen induced mutations and spontaneously (not due to the mutagen) occurring mutations.

All scripts and documented methods are within the subdirectories in this repository.
