# Barley_Mutated

This repository reports the scripts and code written in support of a manuscript by Chaochih Liu et al. examining the nature of mutations generated by sodium azide treatment in barley. Induced variants (SNPs and indels) are compared to naturally occurring variants in barley "genetic resource" lines.

The primary data is derived from DNA resequencing, including Illumina paired-end reads, Illumina reads with a 10X Genomics linked read library preparation, and Oxford Nanopore long reads.

## Analysis overview

Variant calling in this project makes use of GATK 4 to identify SNPs and indels in Illumina resequencing data. Structural variants in Illumina resequencing were identified using multiple lines of evidence through the implementation in the [smoove](https://github.com/brentp/smoove), with primary variant calling in [LUMPY](https://doi.org/10.1186/gb-2014-15-6-r84).

The 10X Genomics linked reads were processed using [LongRanger](https://doi.org/10.1101/gr.234443.118). Oxford Nanopore sequencing was used for read-based verification of structural variants, with variant calls based on [Sniffles](https://doi.org/10.1038/s41592-018-0001-7) as implemented through [Vulcan](https://doi.org/10.1093/gigascience/giab063).

All scripts and detailed methods are within the subdirectories in this repository.
